File: app/src/main/java/com/rafamatias/nlp/presentation/view/activity/TweetsActivity.java
package com.rafamatias.nlp.presentation.view.activity;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.DialogInterface;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;

import com.rafamatias.nlp.R;
import com.rafamatias.nlp.databinding.ActivityTweetsBinding;
import com.rafamatias.nlp.domain.Resource;
import com.rafamatias.nlp.presentation.model.TweetModel;
import com.rafamatias.nlp.presentation.view.adapter.TweetsAdapter;
import com.rafamatias.nlp.presentation.view.fragment.TweetFragment;
import com.rafamatias.nlp.presentation.viewModel.TweetsViewModel;

import java.util.ArrayList;
import java.util.List;

/**
 * A simple {@link AppCompatActivity} subclass.
 * Uses {@link ViewPager} to show fragments of {@link TweetFragment}
 * @see TweetsAdapter
 * @see TweetFragment
 */
public class TweetsActivity extends AppCompatActivity {

    private ActivityTweetsBinding binding;
    private TweetsAdapter viewPagerAdapter;
    private List<TweetModel> viewPagerData = new ArrayList<>();
    private TweetsViewModel viewModel;

    private EditText usernameText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_tweets);

        usernameText = new EditText(this);

        setupViewModel();
        setupViewPager();
    }

    public void actionRetry(View view){
        viewModel.getTweets();
    }

    public void onChangeUsername(View view){
        alertDialogBuilder().create().show();
    }

    private void setupViewModel() {
        viewModel = ViewModelProviders.of(this).get(TweetsViewModel.class);
        viewModel.getTweets().observe(this, getTweetsObserver());
        viewModel.getUsername().observe(this, getUsernameObserver());
    }

    private void setupViewPager() {
        viewPagerAdapter = new TweetsAdapter(getSupportFragmentManager(), viewPagerData);
        binding.viewPager.setAdapter(viewPagerAdapter);
    }

    private Observer<String> getUsernameObserver(){
        return new Observer<String>() {
            @Override
            public void onChanged(@Nullable String resource) {
                binding.textName.setText(getString(R.string.text_username, resource));
            }
        };
    }

    private Observer<Resource<List<TweetModel>>> getTweetsObserver() {
        return new Observer<Resource<List<TweetModel>>>() {
            @Override
            public void onChanged(Resource<List<TweetModel>> resource) {
                switch (resource.state){
                    case ERROR:
                        showError();
                        break;
                    case LOADING:
                        showLoading();
                        break;
                    default:
                        showTweets(resource.data);
                        break;
                }
            }
        };
    }

    public AlertDialog.Builder alertDialogBuilder(){
        LayoutInflater li = LayoutInflater.from(this);
        final View promptView = li.inflate(R.layout.dialog_prompt, null);
        final EditText usernameText = promptView.findViewById(R.id.textUsername);

        return new AlertDialog.Builder(this)
                .setTitle(R.string.title_dialog_change_username)
                .setView(promptView)
                .setPositiveButton(R.string.abc_change, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        viewModel.getTweets(usernameText.getText().toString());
                    }
                });
    }

    private void showLoading() {
        binding.progressBar.setVisibility(View.VISIBLE);
        binding.viewPager.setVisibility(View.GONE);
        binding.containerRetry.setVisibility(View.GONE);
    }

    private void showTweets(List<TweetModel> tweets){
        binding.containerRetry.setVisibility(View.GONE);
        binding.progressBar.setVisibility(View.GONE);
        binding.viewPager.setVisibility(View.VISIBLE);

        tweets.get(0).createdAt();

        viewPagerData.clear();
        viewPagerData.addAll(tweets);
        viewPagerAdapter.notifyDataSetChanged();
    }

    private void showError(){
        binding.containerRetry.setVisibility(View.VISIBLE);
        binding.progressBar.setVisibility(View.GONE);
        binding.viewPager.setVisibility(View.GONE);
    }

}

File: app/src/main/java/br/com/fornaro/tweetssentiment/view/tweets/TweetsActivity.kt
package br.com.fornaro.tweetssentiment.view.tweets

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.databinding.DataBindingUtil
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.widget.Toast
import br.com.fornaro.tweetssentiment.R
import br.com.fornaro.tweetssentiment.common.AppConstants
import br.com.fornaro.tweetssentiment.databinding.ActivityTweetsBinding
import br.com.fornaro.tweetssentiment.model.Tweet
import br.com.fornaro.tweetssentiment.view.main.MainActivity
import br.com.fornaro.tweetssentiment.viewmodel.TweetsViewModel
import kotlinx.android.synthetic.main.activity_tweets.*
import kotlinx.android.synthetic.main.content_scrolling.*

class TweetsActivity : AppCompatActivity(), TweetsAdapter.OnTweetListener {

    private lateinit var binding: ActivityTweetsBinding
    private lateinit var viewModel: TweetsViewModel
    private val viewAdapter = TweetsAdapter(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_tweets)

        val username = intent.extras?.getString(AppConstants.EXTRA_USERNAME)
                ?: throw Exception("extra_username is needed to be put on extras")

        setupToolbar(username)
        setupRecyclerView()
        setupViewModel(username)
    }

    private fun setupToolbar(username: String) {
        toolbar.title = "@$username"
        toolbar.inflateMenu(R.menu.menu_tweets)
        toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.action_select_another_user -> {
                    selectAnotherUser()
                    true
                }
                else -> super.onOptionsItemSelected(it)
            }
        }
    }

    private fun selectAnotherUser() {
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    private fun setupRecyclerView() {
        recyclerView.apply {
            setHasFixedSize(true)
            addItemDecoration(DividerItemDecoration(this@TweetsActivity, DividerItemDecoration.VERTICAL))
            adapter = viewAdapter
        }
    }

    private fun setupViewModel(username: String) {
        viewModel = ViewModelProviders.of(this).get(TweetsViewModel::class.java)
        viewModel.callback = callback

        binding.userInProgress = true
        viewModel.getUser(username).observe(this, Observer {
            binding.userInProgress = false
            binding.resource = it
        })

        binding.tweetsInProgress = true
        viewModel.getTweets(username).observe(this, Observer {
            binding.tweetsInProgress = false
            if (it != null) {
                binding.isEmpty = it.isEmpty()
                viewAdapter.setData(it)
            }
        })
    }

    override fun analyze(tweet: Tweet) {
        viewModel.analyze(tweet)?.observe(this, Observer {
            viewAdapter.analyzedFinished(tweet)
        })
    }

    private val callback = object : TweetsCallback {
        override fun onNoInternetConnection() {
            viewAdapter.notifyDataSetChanged()
            Toast.makeText(this@TweetsActivity, R.string.no_internet, Toast.LENGTH_SHORT).show()
        }
    }
}
